File: core/src/main/java/org/zenika/core/Battle.java
package org.zenika.core;

public class Battle {

    protected FightingPokemon pokemonA;

    protected FightingPokemon pokemonB;

    protected Pokemon winner;

    protected int round;

    public Battle() {
    }

    public Battle(FightingPokemon pokemonA, FightingPokemon pokemonB) {
        this.pokemonA = pokemonA;
        this.pokemonB = pokemonB;
        this.round = 0;
    }

    public FightingPokemon getStarter() {
        return pokemonA.getFightOrder() <= pokemonB.getFightOrder() ? pokemonA : pokemonB;
    }

    public FightingPokemon getEnder() {
        return getStarter() == pokemonA ? pokemonB : pokemonA;
    }

    public BattleStatus getBattleStatus() {
        if (round == 0) {
            return BattleStatus.TO_START;
        }
        if (winner == null) {
            return BattleStatus.PENDING;
        }
        return BattleStatus.ENDED;
    }

    public FightingPokemon getPokemonA() {
        return pokemonA;
    }

    public void setPokemonA(FightingPokemon pokemonA) {
        this.pokemonA = pokemonA;
    }

    public FightingPokemon getPokemonB() {
        return pokemonB;
    }

    public void setPokemonB(FightingPokemon pokemonB) {
        this.pokemonB = pokemonB;
    }

    public Pokemon getWinner() {
        return winner;
    }

    public void setWinner(Pokemon winner) {
        this.winner = winner;
    }

    public int getRound() {
        return round;
    }

    public void setRound(int round) {
        this.round = round;
    }
}

File: core/src/main/kotlin/org/zenika/core/Battle.kt
package org.zenika.core

open class Battle(var pokemonA: FightingPokemon, var pokemonB: FightingPokemon) {
    var winner: Pokemon? = null

    var round: Int = 0

    val starter: FightingPokemon
        get() = if (pokemonA.fightOrder <= pokemonB.fightOrder) pokemonA else pokemonB

    val ender: FightingPokemon
        get() = if (starter == pokemonA) pokemonB else pokemonA

    val battleStatus: BattleStatus
            get() = when {
                round == 0 -> BattleStatus.TO_START
                winner == null -> BattleStatus.PENDING
                else -> BattleStatus.ENDED
            }

}

